src/adapters/curve/products/pool/curvePoolAdapter.ts

```diff
@@ -133,15 +171,46 @@ export class CurvePoolAdapter
     return protocolToken
   }
 
-  /**
-   * Update me.
-   * Add logic that finds the underlying token rates for 1 protocol token
-   */
   protected async getUnderlyingTokenConversionRate(
-    _protocolTokenMetadata: Erc20Metadata,
-    _blockNumber?: number | undefined,
+    protocolTokenMetadata: Erc20Metadata,
+    blockNumber: number | undefined,
   ): Promise<UnderlyingTokenRate[]> {
-    throw new NotImplementedError()
+    const { poolAddress, underlyingTokens, protocolToken } =
+      await this.fetchPoolMetadata(protocolTokenMetadata.address)
+
+    const metaRegistryContract = MetaRegistry__factory.connect(
+      CURVE_META_REGISTRY_CONTRACT,
+      this.provider,
+    )
+
+    const balances = await metaRegistryContract['get_balances(address)'](
+      poolAddress,
+      { blockTag: blockNumber },
+    )
+
+    const lpTokenContract = Erc20__factory.connect(
+      protocolToken.address,
+      this.provider,
+    )
+
+    const supply = await lpTokenContract.totalSupply({ blockTag: blockNumber })
+
+    return balances.slice(0, underlyingTokens.length).map((balance, index) => {
```

Comment:
I'm failing to see the logic behind slicing the array here.

Reply:
ok probably: no dynamic array limitation of vyper so they set the response of the balances to an array which does not match with the amount of underlying tokens, nonetheless Ive switched the implementation.

Reply:
Ill iterate over the underlying token array instead